driver/util.c

```diff
@@ -314,72 +289,59 @@ WnbdProcessDeviceThreadRequests(_In_ PSCSI_DEVICE_INFORMATION DeviceInformation)
-                    StorPortNotification(RequestComplete,
-                                         Element->DeviceExtension,
-                                         Element->Srb);
-                    ExFreePool(Element);
-                    continue;
-            }
+        if(!ValidateScsiRequest(DeviceInformation, Element)) {
+            Element->Srb->DataTransferLength = 0;
+            Element->Srb->SrbStatus = SRB_STATUS_INVALID_REQUEST;
+            StorPortNotification(RequestComplete,
+                                 Element->DeviceExtension,
+                                 Element->Srb);
+            ExFreePool(Element);
+            InterlockedDecrement64(&DeviceInformation->Stats.UnsubmittedIORequests);
+            continue;
         }
 
-        switch (Cdb->AsByte[0]) {
-        case SCSIOP_READ6:
-        case SCSIOP_READ:
-        case SCSIOP_READ12:
-        case SCSIOP_READ16:
-            KeWaitForSingleObject(&DeviceInformation->RequestSemaphore, Executive, KernelMode, FALSE, NULL);
-            if(DeviceInformation->SoftTerminateDevice || DeviceInformation->HardTerminateDevice) {
+        DWORD NbdTransmissionFlags = 0;
+        switch (NbdReqType) {
+        case NBD_CMD_WRITE:
+        case NBD_CMD_TRIM:
+            if (Element->FUA && CHECK_NBD_SEND_FUA(DeviceInformation->UserEntry->NbdFlags)) {
+                NbdTransmissionFlags |= NBD_CMD_FLAG_FUA;
+            }
+        case NBD_CMD_READ:
+        case NBD_CMD_FLUSH:
+            KeWaitForSingleObject(
+                &DeviceInformation->RequestSemaphore,
+                Executive, KernelMode, FALSE, NULL);
+            if(DeviceInformation->SoftTerminateDevice ||
+                    DeviceInformation->HardTerminateDevice) {
                 return;
             }
             ExInterlockedInsertTailList(
                 &DeviceInformation->ReplyListHead,
                 &Element->Link, &DeviceInformation->ReplyListLock);
-            WNBD_LOG_LOUD("Pending request. Address: %p Tag: 0x%llx",
-                          Element->Srb, Element->Tag);
-            Status = WnbdProcessDeviceThreadRequestsReads(DeviceInformation, Element);
-            break;
+            WNBD_LOG_LOUD("Sending %s request. Address: %p Tag: 0x%llx. FUA: %d",
+                          NbdRequestTypeStr(NbdReqType), Element->Srb, Element->Tag,
+                          Element->FUA);
 
-        case SCSIOP_WRITE6:
-        case SCSIOP_WRITE:
-        case SCSIOP_WRITE12:
-        case SCSIOP_WRITE16:
-            KeWaitForSingleObject(&DeviceInformation->RequestSemaphore, Executive, KernelMode, FALSE, NULL);
-            if(DeviceInformation->SoftTerminateDevice || DeviceInformation->HardTerminateDevice) {
-                return;
+            if(NbdReqType == NBD_CMD_WRITE){
+                Status = WnbdRequestWrite(DeviceInformation, Element,
+                                          NbdTransmissionFlags);
+            }
+            else {
```

Comment:
We don't have a proper coding style documented. Can you please use the form:
if () {
} else {
}
It will be consistent we what we have already.

Reply:
Will do. Thanks for mentioning it.